File: Codex/Python/2025/Day7_Solution_2025.py
def main() -> None:
    with open("Day7_input.txt", "r", encoding="utf-8") as file:
        grid = [line.rstrip("\n") for line in file if line.rstrip("\n")]

    start_col = grid[0].index("S")
    active = {start_col}
    split_count = 0
    width = len(grid[0])

    for row in grid[1:]:
        next_active = set()
        for col in active:
            if row[col] == "^":
                split_count += 1
                if col > 0:
                    next_active.add(col - 1)
                if col + 1 < width:
                    next_active.add(col + 1)
            else:
                next_active.add(col)
        active = next_active
        if not active:
            break

    print(split_count)


if __name__ == "__main__":
    main()

File: Codex/Rust/2025/Day7_Solution_2025.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("Day7_input.txt").expect("failed to read Day7_input.txt");
    let lines: Vec<&str> = input.lines().collect();
    if lines.is_empty() {
        println!("0");
        return;
    }

    let width = lines[0].len();
    let grid: Vec<&[u8]> = lines.iter().map(|line| line.as_bytes()).collect();

    let start_col = grid[0]
        .iter()
        .position(|&cell| cell == b'S')
        .expect("missing start position");

    let mut current = vec![false; width];
    current[start_col] = true;

    let mut splits: u64 = 0;

    for row in &grid[1..] {
        let mut next = vec![false; width];

        for col in 0..width {
            if !current[col] {
                continue;
            }

            match row[col] {
                b'^' => {
                    splits += 1;
                    if col > 0 {
                        next[col - 1] = true;
                    }
                    if col + 1 < width {
                        next[col + 1] = true;
                    }
                }
                b'.' | b'S' => {
                    next[col] = true;
                }
                _ => panic!("unexpected cell"),
            }
        }

        current = next;
    }

    println!("{splits}");
}
